Sources/Algorithms/Searchers/Searcher.swift

```diff
@@ -0,0 +1,77 @@
+public protocol CollectionSearcher {
+  associatedtype Searched: Collection
+  associatedtype State = ()
+  func initialState(_ searched: Searched) -> State
+  func search(
+    _ searched: Searched,
+    subrange: Range<Searched.Index>,
+    _ state: inout State
+  ) -> Range<Searched.Index>?
```

Comment:
An interesting challenge will be whether or how we guarantee that this will be invoked. E.g. State might be specific to the last index returned, so we may want to ensure we're invoking it monotonically.

It's unfortunate that access control with protocols is so coarse-grained. Ideally, one could make a public conformer without making these low-level interfaces public. Essentially, all that is really required at an ABI level is `@_usableFromInline`.

Elsewhere we have underscored the name for some of these things, but hopefully we can avoid forcing that pattern.